File: packages/membership-sdk/src/types/plan.ts
import type { PublicKey } from "@solana/web3.js";

/**
 * On-chain Plan structure
 */
export type PlanAccount = {
  /** Plan ID (same as PDA public key) */
  planId: PublicKey;
  /** Business that owns this plan */
  business: PublicKey;
  /** Encrypted plan name (32 bytes) */
  encryptedName: Uint8Array;
  /** Encrypted plan description (64 bytes) */
  encryptedDescription: Uint8Array;
  /** Price in USDC (6 decimals) */
  priceUsdc: bigint;
  /** Billing cycle in seconds */
  billingCycleSeconds: number;
  /** Unix timestamp when created */
  createdAt: bigint;
  /** Whether the plan is active */
  isActive: boolean;
  /** Encrypted subscription count */
  encryptedSubscriptionCount: Uint8Array;
  /** Encryption nonce */
  nonce: bigint;
  /** Sequential plan number for this business */
  planNonce: bigint;
  /** PDA bump seed */
  bump: number;
};

/**
 * Input for creating a new plan
 */
export type CreatePlanInput = {
  /** Plan name (will be encrypted) */
  name: string;
  /** Plan description (will be encrypted) */
  description: string;
  /** Price in USDC (human readable, e.g. 9.99) */
  priceUsdc: number;
  /** Billing cycle in days */
  billingCycleDays: number;
};

/**
 * Plan with decrypted metadata
 */
export type Plan = {
  /** Public key of the plan account PDA */
  publicKey: PublicKey;
  /** Business that owns this plan */
  business: PublicKey;
  /** Plan name (decrypted if authorized) */
  name?: string;
  /** Plan description (decrypted if authorized) */
  description?: string;
  /** Price in USDC (human readable) */
  priceUsdc: number;
  /** Billing cycle in days */
  billingCycleDays: number;
  /** Unix timestamp when created */
  createdAt: Date;
  /** Whether the plan is active */
  isActive: boolean;
  /** Subscription count (decrypted if authorized) */
  subscriptionCount?: number;
};

/**
 * Plan list filter options
 */
export type PlanFilter = {
  /** Filter by business */
  business?: PublicKey;
  /** Filter by active status */
  isActive?: boolean;
  /** Maximum number of results */
  limit?: number;
  /** Offset for pagination */
  offset?: number;
};

File: subly_devnet/programs/subly_devnet/src/state/plan.rs
use anchor_lang::prelude::*;

/// Plan - stores subscription plan information
/// PDA seeds: ["plan", business, plan_nonce]
#[account]
pub struct Plan {
    /// Plan's own address (for reference)
    pub plan_id: Pubkey,
    /// Reference to business account
    pub business: Pubkey,
    /// Encrypted plan name (Arcium encrypted)
    pub encrypted_name: [u8; 32],
    /// Encrypted plan description (Arcium encrypted)
    pub encrypted_description: [u8; 64],
    /// Price in USDC (6 decimals, e.g., 10_000_000 = 10 USDC)
    pub price_usdc: u64,
    /// Billing cycle in seconds
    pub billing_cycle_seconds: u32,
    /// Plan creation timestamp
    pub created_at: i64,
    /// Whether the plan is active
    pub is_active: bool,
    /// Encrypted subscription count (Arcium encrypted, only count is known)
    pub encrypted_subscription_count: [u8; 32],
    /// Nonce for encryption
    pub nonce: u128,
    /// Plan nonce (sequential number for this business)
    pub plan_nonce: u64,
    /// PDA bump
    pub bump: u8,
    /// Whether count encryption is pending (waiting for Arcium callback)
    pub pending_count_encryption: bool,
}

impl Plan {
    /// Calculate account space
    pub const SPACE: usize = 8    // discriminator
        + 32                       // plan_id: Pubkey
        + 32                       // business: Pubkey
        + 32                       // encrypted_name: [u8; 32]
        + 64                       // encrypted_description: [u8; 64]
        + 8                        // price_usdc: u64
        + 4                        // billing_cycle_seconds: u32
        + 8                        // created_at: i64
        + 1                        // is_active: bool
        + 32                       // encrypted_subscription_count: [u8; 32]
        + 16                       // nonce: u128
        + 8                        // plan_nonce: u64
        + 1                        // bump: u8
        + 1;                       // pending_count_encryption: bool
    // Total: 247 bytes
}
